discovery/discovery.go

```diff
@@ -10,23 +14,92 @@ import (
 	"github.com/golang/glog"
 )
 
-type offchainOrchestrator struct {
-	uri   *url.URL
+const GetOrchestratorsTimeoutLoop = 3 * time.Second
+
+type offchainOrchestrators struct {
+	uri   []*url.URL
 	bcast server.Broadcaster
 }
 
-func NewOffchainOrchestrator(node *core.LivepeerNode, address string) *offchainOrchestrator {
-	addr := "https://" + address
-	uri, err := url.ParseRequestURI(addr)
+func NewOffchainOrchestrator(node *core.LivepeerNode, addresses []string) *offchainOrchestrators {
+	var uris []*url.URL
+
+	for _, addr := range addresses {
+		if !strings.HasPrefix(addr, "http") {
+			addr = "https://" + addr
+		}
+		uri, err := url.ParseRequestURI(addr)
+		if err != nil {
+			glog.Error("Could not parse orchestrator URI: ", err)
+			continue
+		}
+		uris = append(uris, uri)
+	}
+
+	if len(uris) <= 0 {
+		glog.Error("Could not parse orchAddresses given - no URIs returned ")
+	}
+
+	bcast := core.NewBroadcaster(node)
+	return &offchainOrchestrators{bcast: bcast, uri: uris}
+}
+
+func NewOffchainOrchestratorFromOnchainList(node *core.LivepeerNode) *offchainOrchestrators {
+	// if livepeer running in offchain mode, return nil
+	if node.Eth == nil {
+		return nil
+	}
+
+	orchestrators, err := node.Eth.RegisteredTranscoders()
 	if err != nil {
-		glog.Error("Could not parse orchestrator URI: ", err)
+		glog.Error(err)
 		return nil
 	}
-	bcast := core.NewBroadcaster(node)
-	return &offchainOrchestrator{bcast: bcast, uri: uri}
+
+	var addresses []string
+	for _, orch := range orchestrators {
+		addresses = append(addresses, orch.ServiceURI)
+	}
+
+	return NewOffchainOrchestrator(node, addresses)
 }
 
-func (o *offchainOrchestrator) GetOrchestrators(numOrchestrators int) ([]*net.OrchestratorInfo, error) {
-	tinfo, err := server.GetOrchestratorInfo(o.bcast, o.uri)
-	return []*net.OrchestratorInfo{tinfo}, err
+func (o *offchainOrchestrators) GetOrchestrators(numOrchestrators int) ([]*net.OrchestratorInfo, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), GetOrchestratorsTimeoutLoop)
+	orchInfos := []*net.OrchestratorInfo{}
+	orchChan := make(chan struct{})
+	numResp := 0
+	numSuccessResp := 0
+	respLock := sync.Mutex{}
+	getOrchInfo := func(uri *url.URL) {
+
+		info, err := server.GetOrchestratorInfo(ctx, o.bcast, uri)
+		if err != nil {
+			glog.Errorf("Error getting OrchestratorInfo for uri: %v", uri)
+		}
+		respLock.Lock()
+		defer respLock.Unlock()
+		numResp++
+		orchInfos = append(orchInfos, info)
+		numSuccessResp++
```

Comment:
orchInfo should only be appended to and `numSucessResp` incremented if the response was actually successful (`err` non-nil)

Reply:
ah yessss. Missing a return after line 78. I'm moving `numResp++` up to before the error check after GetOrchestratorInfo